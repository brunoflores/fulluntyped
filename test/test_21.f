Counter = { get: Unit -> Nat, inc: Unit -> Unit };

-- The "representation type" of the "class" below.
CounterRep = { x: Ref Nat };

-- Document in code that this record must type-check as a Counter.
counterClass =
  lambda r: CounterRep.
    { get = lambda _: Unit. !(r.x),
      inc = lambda _: Unit. r.x := succ (!(r.x)) } as Counter;

-- Instantiate a counter using a selected implementation.
newCounter =
  lambda _: Unit. let r = { x = ref 1 } in
    counterClass r;

c = newCounter unit;
