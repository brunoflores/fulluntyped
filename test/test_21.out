Counter: { get: Unit->Nat, inc: Unit->Unit }
CounterRep: { x: Ref Nat }
counterClass: CounterRep -> Counter
newCounter: Unit -> Counter
c: Counter
